test/cql-pytest/test_unset.py

```diff
@@ -0,0 +1,167 @@
+    i = 4
+    for unset_a in [False, True]:
+        for unset_b in [False, True]:
+            for unset_c in [False, True]:
+                if unset_a:
+                    newa = UNSET_VALUE
+                else:
+                    newa = i
+                    a = i
+                    i += 1
+                if unset_b:
+                    newb = UNSET_VALUE
+                else:
+                    newb = i
+                    b = i
+                    i += 1
+                if unset_c:
+                    newc = UNSET_VALUE
+                else:
+                    newc = i
+                    c = i
+                    i += 1
+                cql.execute(stmt, [newa, newb, newc])
+                assert [(a, b, c)] == list(cql.execute(f'SELECT a,b,c FROM {table1} WHERE p = {p}'))
+
+# The expression "SET a=?" is skipped if the bound value is UNSET_VALUE.
+# But what if it is a more complex expression, like "SET li=li+?" (list append)
+# or "SET a=(int)?+1" (arithmetic expression on the bind variable)? Does
+# the SET also get skipped? Unfortunately, the answer is no: Cassandra
+# refuses or fails to evaluate expressions involving an UNSET_VALUE, and
+# it causes the whole write request to fail instead of parts of it being
+# skipped. Noticing this, we decided (see discussion in pull request #12517)
+# that this will become a rule: expressions involving UNSET_VALUE will
+# always be refused. The following tests verify this in a few cases.
+
+@pytest.mark.xfail(reason="append with unset doesn't produce error")
+def test_update_unset_value_expr_1(cql, table1):
+    p = unique_key_int()
+    stmt = cql.prepare(f'UPDATE {table1} SET li=li+? WHERE p={p}')
+    cql.execute(stmt, [[7]])
+    assert [([7],)] == list(cql.execute(f'SELECT li FROM {table1} WHERE p = {p}'))
+    # Ideally, trying to use UNSET_VALUE in an expression where it is not
+    # allowed should result in a clean InvalidRequest. But currently in
+    # Cassandra it results in an internal casting error and NoHostAvailable.
+    with pytest.raises((NoHostAvailable, InvalidRequest)):
+        cql.execute(stmt, [UNSET_VALUE])
+
+@pytest.mark.xfail(reason="issue #2693")
+def test_update_unset_value_expr_2(cql, table1):
+    p = unique_key_int()
+    stmt = cql.prepare(f'UPDATE {table1} SET a=(int)?+1 WHERE p={p}')
+    cql.execute(stmt, [7])
+    assert [(8,)] == list(cql.execute(f'SELECT a FROM {table1} WHERE p = {p}'))
+    with pytest.raises(InvalidRequest):
+        cql.execute(stmt, [UNSET_VALUE])
+
+# Despite the above examples of expressions and expression-like syntaxes
+# that do not allow UNSET_VALUE, the CQL v4 documentation explicitly mentions
+# that "an unset bind counter operation does not change the counter value.".
+# So "c = c + ?" for a counter, when given an UNSET_VALUE, will causes the
+# write to be skipped, without error.
+def test_unset_counter_increment(cql, test_keyspace):
```

Comment:
Another example of a primitive operation expressed as a complex expression.

I guess we could change unset to "if any unset variables are present in the expression tree, the sub-operation is skipped, but I don't think it's worth the effort.

Reply:
Right. What you said that "we could" was my original thinking, but you convinced me that it's not worth the effort, so I'm writing tests that will explain (and enshrine) the decision we reached, so it won't be suddenly (and partially) reverted in some future patch of someone that isn't familiar with the history of this decision.

Reply:
btw, we do have a user of unset - sstableloader. It uses a few prepared statements to push any combination of mutations, even if the schema has many columns.